src/packages/hex.gleam

```diff
@@ -3,41 +3,204 @@ import gleam/json
 import gleam/http
 import gleam/result
 import gleam/hackney
+import gleam/list
 import gleam/dynamic.{Decoder} as d
+import time/parse.{parse_iso8601_to_gregorian_seconds}
+import gleam/pgo
+import gleam/string
+import gleam/int
+import gleam/option.{Option}
 
 pub type Error {
   HackneyError(hackney.Error)
   JsonError(json.DecodeError)
+  DatabaseError(pgo.QueryError)
+  ItemNotInListError(Nil)
 }
 
-pub fn query() {
+pub fn query(db: pgo.Connection) {
+  try last_scanned = get_last_scanned(db)
+
+  try packages = query_all_packages([], 1, last_scanned)
+
+  // Only update last scanned at if packages were actually
+  // found that way we only ever use the time from
+  // Hex's API 
+  case packages
+  |> list.length >= 1 {
+    True -> {
+      let len =
+        packages
+        |> list.length
+      assert Ok(last_package) =
+        packages
+        |> list.at(len - 1)
+        |> result.map_error(ItemNotInListError)
+      assert Ok(_) =
+        update_last_scanned(db, parse.parse_iso8601(last_package.updated_at))
+      Ok(Nil)
+    }
+    _ -> Ok(Nil)
+  }
+
+  packages
+  |> list.filter_map(sort_packages)
+  |> io.debug
+  |> Ok
+}
+
+const last_scanned_query = "SELECT date_trunc('second', scanned_at) FROM previous_hex_api_scan LIMIT 1;"
+
+fn get_last_scanned(db: pgo.Connection) -> Result(Int, Error) {
+  let ints = d.tuple3(d.int, d.int, d.int)
+  let datetime_decoder = d.tuple2(ints, ints)
+
+  try response =
+    pgo.execute(last_scanned_query, db, [], d.element(0, datetime_decoder))
+    |> result.map_error(DatabaseError)
+
+  case response.rows {
+    [first, ..] -> Ok(parse.to_gregorian_seconds(first))
+    _ -> Ok(0)
+  }
+}
+
+fn update_last_scanned(
+  db: pgo.Connection,
+  date: parse.LocalDateTime,
+) -> Result(pgo.Returned(d.Dynamic), Error) {
+  // It is done this way as I could not figure out how to
+  // inject a time as a string with the built in $1 system but
+  // this is safe as the value is constructed from a known datatype
+  pgo.execute(
+    "UPDATE previous_hex_api_scan SET scanned_at = '"
+    |> string.append(parse.to_pg_time(date))
+    |> string.append("';"),
+    db,
+    [],
+    d.dynamic,
+  )
+  |> result.map_error(DatabaseError)
+}
+
+fn sort_packages(package: Package) -> Result(Package, Error) {
+  // TODO Sort Gleam Packages
+  // 1. Check if already in DB
+  // 2. Check latest release
+  //    If is gleam: keep, else: remove
+  // 3. Save new release in DB
+  // 4. Update package in DB if out of date
+  // 5. Return the new list of DB Package Type
+  try latest_release =
+    package.releases
+    |> list.at(0)
+    |> result.map_error(ItemNotInListError)
```

Comment:
I don't think we should ever have an error for if a package has no release, instead there is nothing to do for this package.

Reply:
Fixed by another fix

Reply:
This hasn't been done yet

Reply:
Fixed this now so it gives a more descriptive error that is compliant with the doc comment for the function